src/main/java/io/sqreen/powerwaf/PowerwafContext.java

```diff
@@ -123,11 +124,23 @@ public Powerwaf.ResultWithData runRules(Map<String, Object> parameters,
                 // serialization could be extracted out of the lock
                 ByteBufferSerializer serializer = new ByteBufferSerializer(limits);
                 long before = System.nanoTime();
-                ByteBufferSerializer.ArenaLease lease;
-                try {
-                    lease = serializer.serialize(parameters);
-                } catch (Exception e) {
-                    throw new RuntimeException("Exception encoding parameters", e);
+
+                ByteBuffer persistentBuffer = null;
+                if (persistentData != null) {
+                    try (ByteBufferSerializer.ArenaLease lease = serializer.serialize(persistentData)) {
+                        persistentBuffer = lease.getFirstPWArgsByteBuffer();
+                    } catch (Exception e) {
+                        throw new RuntimeException("Exception encoding parameters", e);
+                    }
+                }
+
+                ByteBuffer ephemeralBuffer = null;
+                if (ephemeralData != null) {
+                    try (ByteBufferSerializer.ArenaLease lease = serializer.serialize(ephemeralData)) {
+                        ephemeralBuffer = lease.getFirstPWArgsByteBuffer();
+                    } catch (Exception e) {
+                        throw new RuntimeException("Exception encoding parameters", e);
```

Comment:
It's a possibility, but it would be better IMO to use the same arena. Of course, then `getFirstPWArgsByteBuffer()` couldn't be used, you'd have to take a slice of the ByteBuffer with the first PWArgs of the ephermal.

Ultimately though, this method could also be removed, because now we're only using the additive interface.

Reply:
I'll remove this method in next PR